File: backend/services/cards_service.py
from __future__ import annotations

from typing import Any

from ..cards.logic import answer_card as answer_card_logic, next_card as next_card_logic, start_cards_session as start_cards_session_logic


def start_cards_session(*, user_id: str, domain: str, content_type: str | None, profession: str | None, level: str | None, adaptive: bool = False, limit: int = 10) -> dict[str, Any]:
    return start_cards_session_logic(
        user_id=user_id,
        domain=domain,
        content_type=content_type,
        profession=profession,
        level=level,
        adaptive=adaptive,
        limit=limit,
    )


def next_card(*, user_id: str, session_id: str) -> dict[str, Any]:
    return next_card_logic(user_id=user_id, session_id=session_id)


def answer_card(*, user_id: str, session_id: str, user_answer: str) -> dict[str, Any]:
    return answer_card_logic(user_id=user_id, session_id=session_id, user_answer=user_answer)

File: frontend/app/services/cardsService.ts
import { apiRequest } from "./apiClient";
import type { ApiEnvelope } from "../state/types";

export async function startCardsSession(payload: {
  domain: string;
  content_type: string | null;
  profession: string | null;
  level: string | null;
}): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: "/api/v1/cards/session/start",
    auth: true,
    body: payload,
  });
}

export async function startAdaptiveCardsSession(params: {
  domain: string;
  content_type?: string;
  profession?: string;
  level?: string;
  limit?: number;
}): Promise<ApiEnvelope<any>> {
  const query = new URLSearchParams();
  query.set("domain", params.domain);
  if (params.content_type) query.set("content_type", params.content_type);
  if (params.profession) query.set("profession", params.profession);
  if (params.level) query.set("level", params.level);
  if (params.limit) query.set("limit", String(params.limit));
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/cards/session/adaptive/start?${query.toString()}`,
    auth: true,
  });
}

export async function fetchNextCard(sessionId: string): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "GET",
    path: `/api/v1/cards/session/${sessionId}/next`,
    auth: true,
  });
}

export async function submitCardAnswer(sessionId: string, payload: { user_answer: string }): Promise<ApiEnvelope<any>> {
  return apiRequest<any>({
    method: "POST",
    path: `/api/v1/cards/session/${sessionId}/answer`,
    auth: true,
    body: payload,
  });
}
